File: Topic-Based/Tree/199. Binary Tree Right Side View/Solution-DFS.java
// Definition for a binary tree node.
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();

        rightSide(root, result, 0);
        return result;
    }

    public void rightSide(TreeNode root, List<Integer> result, int depth) {
        // Using preorder traversal with changes of NRL i.e first executing right child
        // as we need to find the rightmost value

        if (root == null)
            return;
        if (depth >= result.size())
            result.add(root.val);
        if (root.right != null)
            rightSide(root.right, result, depth + 1);
        if (root.left != null)
            rightSide(root.left, result, depth + 1);
    }
}

File: Monthly-Challenges/May-2022/Day 15 - 1302. Deepest Leaves Sum/Solution-DFS.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int deepestLeavesSum(TreeNode *root)
    {
        if (!root)
            return 0;
        int depth = 1;
        dfs(root, 1, depth);
        int sum = 0;
        findSum(root, 1, depth, sum);

        return sum;
    }

    void dfs(TreeNode *node, int level, int &depth)
    {
        depth = max(depth, level);

        if (node->left)
            dfs(node->left, level + 1, depth);
        if (node->right)
            dfs(node->right, level + 1, depth);
    }

    void findSum(TreeNode *node, int level, int depth, int &sum)
    {
        if (level == depth)
            sum += node->val;

        if (node->left)
            findSum(node->left, level + 1, depth, sum);
        if (node->right)
            findSum(node->right, level + 1, depth, sum);
    }
};
